### PR TITLE
Enable OOT credential provider for all CI templates

### LINKED ISSUE #4110: Support OOT credential provider for all templates
/kind bug

[Before submitting an issue, have you checked the [Troubleshooting Guide](https://capz.sigs.k8s.io/topics/troubleshooting.html)?]

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]
With latest k8s (1.29+), in-tree credential provider is disabled by default and current capz templates don't support OOT one.

**What did you expect to happen:**
All capz templates support OOT credential provider.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]
I have updated main branch ci-entrypoint to support OOT credential provider. Also for cloud-provider-azure, I have created a template for OOT and it works fine.
https://raw.githubusercontent.com/kubernetes-sigs/cloud-provider-azure/master/tests/k8s-azure/manifest/cluster-api/cluster-template-prow-ci-version-oot-credential-provider.yaml

Key change in template:
Scripts to download credential provider binary and config file
```
# linux
        echo "Use OOT credential provider"
        mkdir -p /var/lib/kubelet/credential-provider
        # Update the link below
        curl --retry 10 --retry-delay 5 -Lo /var/lib/kubelet/credential-provider/acr-credential-provider "https://${AZURE_STORAGE_ACCOUNT}.blob.core.windows.net/${JOB_NAME}/${IMAGE_TAG_ACR_CREDENTIAL_PROVIDER}/azure-acr-credential-provider"
        chmod 755 /var/lib/kubelet/credential-provider/acr-credential-provider

        # Update the link below
        curl --retry 10 --retry-delay 5 -Lo /var/lib/kubelet/credential-provider-config.yaml https://raw.githubusercontent.com/kubernetes-sigs/cloud-provider-azure/master/examples/out-of-tree/credential-provider-config.yaml
        chmod 644 /var/lib/kubelet/credential-provider-config.yaml

# windows
          echo "Use OOT credential provider"
          mkdir C:\var\lib\kubelet\credential-provider
          # Update the link below
          curl.exe --retry 10 --retry-delay 5 -L "https://${AZURE_STORAGE_ACCOUNT}.blob.core.windows.net/${JOB_NAME}/${IMAGE_TAG_ACR_CREDENTIAL_PROVIDER}/azure-acr-credential-provider.exe" --output C:\var\lib\kubelet\credential-provider\acr-credential-provider.exe
          cp C:\var\lib\kubelet\credential-provider\acr-credential-provider.exe C:\var\lib\kubelet\credential-provider\acr-credential-provider

          # Update the link below
          curl.exe --retry 10 --retry-delay 5 -L https://raw.githubusercontent.com/kubernetes-sigs/cloud-provider-azure/master/examples/out-of-tree/credential-provider-config-win.yaml --output C:\var\lib\kubelet\credential-provider-config.yaml
```

kubelet args:
```
        kubeletExtraArgs:
          image-credential-provider-bin-dir: /var/lib/kubelet/credential-provider
          image-credential-provider-config: /var/lib/kubelet/credential-provider-config.yaml
```

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4110 

**Special notes for your reviewer**:

- [x] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: templates/test/ci/prow-ci-version/patches/oot-credential-provider.yaml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      [[ $(id -u) != 0 ]] && SUDO="sudo" || SUDO=""
+
+      echo "Use OOT credential provider"
+      mkdir -p /var/lib/kubelet/credential-provider
+      # Update the link below
+      curl --retry 10 --retry-delay 5 -Lo /var/lib/kubelet/credential-provider/acr-credential-provider "https://${AZURE_STORAGE_ACCOUNT}.blob.core.windows.net/${JOB_NAME}/${IMAGE_TAG_ACR_CREDENTIAL_PROVIDER}/azure-acr-credential-provider"
+      chmod 755 /var/lib/kubelet/credential-provider/acr-credential-provider
+
+      # Update the link below
+      curl --retry 10 --retry-delay 5 -Lo /var/lib/kubelet/credential-provider-config.yaml https://raw.githubusercontent.com/kubernetes-sigs/cloud-provider-azure/master/examples/out-of-tree/credential-provider-config.yaml
```
Comment (CecileRobertMichon) [inline templates/test/ci/prow-ci-version/patches/oot-credential-provider.yaml:19]: TODO: use release branch to avoid pulling from master

**Review by CecileRobertMichon — COMMENTED** [review]
/cc @lzhecheng @mainred

File: templates/test/ci/cluster-template-prow-ci-version-dual-stack.yaml [hunk 2]
```diff
@@ -124,6 +124,26 @@ spec:
         set -o errexit
         [[ $(id -u) != 0 ]] && SUDO="sudo" || SUDO=""
 
+        echo "Use OOT credential provider"
+        mkdir -p /var/lib/kubelet/credential-provider
+        # Update the link below
```
Comment (lzhecheng) [inline templates/test/ci/cluster-template-prow-ci-version-dual-stack.yaml:129]: Actually, this is a comment that should be removed but I forget to in our customized template...

**Review by lzhecheng — COMMENTED** [review]
Overall LGTM

**Review by CecileRobertMichon — COMMENTED** [review]
/hold for squash

File: templates/test/ci/cluster-template-prow-ci-version-dual-stack.yaml (hunk 2 above)
Comment (CecileRobertMichon) [reply templates/test/ci/cluster-template-prow-ci-version-dual-stack.yaml]: removed

**Comment by CecileRobertMichon** [comment]
> Error response from daemon: unauthorized: authentication required

another docker transient issue

/retest

**Review by CecileRobertMichon — COMMENTED** [review]
/hold cancel

this is ready

/assign @mboersma

**Review by willie-yao — APPROVED** [review]
/lgtm

Just for my own understanding, which CI jobs would these apply to? It doesn't seem to touch all CAPZ prow templates. I guess I'm a bit confused about what the ci-version templates are used for

**Comment by CecileRobertMichon** [comment]
@willie-yao this would be all the CAPZ PR jobs with suffix `with-ci-artifacts` + all the periodic jobs testing the various release branches of k8s https://testgrid.k8s.io/provider-azure-master-signal, essentially anything that's testing kubernetes branches instead of stable k8s releases

**Review by CecileRobertMichon — COMMENTED** [review]
/approve

**Pull request merged**